### DISCUSSION TITLE
How to embed an existing language into my DSL and support cross language reference?

### DISCUSSION THREAD

**Discussion by suerta-git** [eclipse-langium/langium#1115] [Q&A] [answered]
Hi team. I am building an in-house DSL for modeling our business. I hope it can inline some expressions written by other language (e.g. JavaScript) for calculation. Here is an example:

```
type HelloWorld {

  number FieldA

  number FieldB = js`Math.pow(FieldA, 3)`  //  Here is a JavaScript expression

}
```

Requirements:

- JavaScript syntax highlight, autocomplete and validation
- Support `go to definition` for cross references in expressions
- Nice to have: Type-based autocomplete and validation for referenced fields

Questions:

- Is it possible to implement above requirements and how?
- Is it possible to build a plugin mechanism to support embedding any language and provide consistent IDE functionalities?
- Can langium provide some apis for language embedding or at least a tutorial?

Thanks a lot!

**Comment by msujew** [comment, marked as answer]
Hey @suerta-git,

embedded programming languages are quite common in vscode, and although we don't have a tutorial on this topic, [they do](https://code.visualstudio.com/api/language-extensions/embedded-languages). I haven't developed an embedded language for vscode yet, but the second approach they show (i.e. [request forwarding](https://code.visualstudio.com/api/language-extensions/embedded-languages#request-forwarding)) looks pretty promising. Both can work, though the other approach seems to be associated with a lot of maintenance effort.

Unrelated to vscode: I'm not sure how much Langium can help in that regard. Embedded programming languages are generally outside the context of the language server protocol. We try to operate within that as much as possible.

**Reply by suerta-git** [reply]
Thanks, the docs you referred are very helpful.

> I'm not sure how much Langium can help in that regard. Embedded programming languages are generally outside the context of the language server protocol. We try to operate within that as much as possible.

I'm not either. 🤣

But if this requirement is common, maybe langium can provide a batch of abstract interfaces let developers can integrate other language with langium-based language more easily. 

Such as supports to mark embedded language regions in langium DSL, be configurable on request forwarding, give options to embed other language services in generated extension and have easy-to-use apis to support cross-language features...

**Reply by msujew** [reply]
> have easy-to-use apis to support cross-language features

Cross language references are already supported pretty well. We recently worked on a language were we cross reference Java code, see [here](https://github.com/BBx-Kitchen/bbj-language-server/). 

> be configurable on request forwarding

Request forwarding happens on the language client side, not the language server side. Not much we can do there.

I'm open to improvements to the dev experience in that regard though. If you have any suggestions, please create a feature request issue for that.
